<a href="https://colab.research.google.com/github/LegendSeyi/Thai_Road_Accident_Data_Analysis/blob/main/Thai_Road_Accident_data_analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **THAI ROAD ACCIDENT ANALYSIS**
#### **_By Adeyemi Oluwaseyi Emmanuel_**

[Click here to check Linkedin profile](https://www.linkedin.com/in/oluwaseyi-adeyemi-33b1ab197/)

[Twitter link](https://twitter.com/AmLegendseyi)

## **INTRODUCTION**

This dataset provides comprehensive statistics on recorded road accidents in Thailand, spanning from approximately 2019 to 2022. The data was sourced from raw information provided by the Office of the Permanent Secretary, Ministry of Transport, which is also utilized in this public dashboard for easier access and visualization. The dataset encompasses various aspects of road accidents and aims to shed light on the trends and patterns within this critical area of concern, analysis of this data could be crucial in guiding road safety policies and measures👍.

We import our required libraries

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
## creating a dictionary in python and converting it to a dataframe
tb = {'Column':['acc_code',
    'incident_datetime',
    'report_datetime',
    'province_th',
    'province_en',
    'agency',
    'route',
    'vehicle_type',
    'presumed_cause',
    'accident_type',
    'number_of_vehicles_involved',
    'number_of_fatalities',
    'number_of_injuries',
    'weather_condition',
    'latitude',
    'longitude',
    'road_description',
    'slope_description'],
    'Description':['The accident code or identifier',
    'The date and time of the accident occurrence',
    'The date and time when the accident was reported',
    'The name of the province in Thailand, written in Thai',
    'The name of the province in Thailand, written in English',
    'The government agency responsible for the road and traffic management',
    'The route or road segment where the accident occurred',
    'The type of vehicle involved in the accident',
    'The presumed cause or reason for the accident',
    'The type or nature of the accident',
    'The number of vehicles involved in the accident',
    'The number of fatalities resulting from the accident',
    'The number of injuries resulting from the accident',
    'The weather condition at the time of the accident',
    'The latitude coordinate of the accident location',
    'The longitude coordinate of the accident location',
    'The description of the road type or configuration where the accident occurred',
    'The description of the slope condition at the accident location']
     }

In [ ]:
## Converting the dictionary to a data frame
dico = pd.DataFrame(tb)

In [ ]:
dico

Column  \
0                      acc_code   
1             incident_datetime   
2               report_datetime   
3                   province_th   
4                   province_en   
5                        agency   
6                         route   
7                  vehicle_type   
8                presumed_cause   
9                 accident_type   
10  number_of_vehicles_involved   
11         number_of_fatalities   
12           number_of_injuries   
13            weather_condition   
14                     latitude   
15                    longitude   
16             road_description   
17            slope_description   

                                          Description  
0                     The accident code or identifier  
1        The date and time of the accident occurrence  
2    The date and time when the accident was reported  
3   The name of the province in Thailand, written ...  
4   The name of the province in Thailand, written ...  
5   The government agency responsible for the road...  
6   The route or road segment where the accident o...  
7        The type of vehicle involved in the accident  
8       The presumed cause or reason for the accident  
9                  The type or nature of the accident  
10    The number of vehicles involved in the accident  
11  The number of fatalities resulting from the ac...  
12  The number of injuries resulting from the acci...  
13  The weather condition at the time of the accident  
14   The latitude coordinate of the accident location  
15  The longitude coordinate of the accident location  
16  The description of the road type or configurat...  
17  The description of the slope condition at the ...

The above table is the describtion of each columns in the dataset....

With this information, we can now import our daraset for the analysis

## **UNDERSTANDING THE DATASET**

In [ ]:
# We will use the pandas library to import our dataset

In [ ]:
# We can import our dataset from github


url = 'https://raw.githubusercontent.com/LegendSeyi/Dataset/main/thai_road_accident_2019_2022.csv'

pd.set_option('display.max_columns', None)
data = pd.read_csv(url, encoding= 'unicode_escape')


In [ ]:
data.rename(columns = {'acc_code': 'acc_code'}, inplace = True)

In [ ]:
data.head()

acc_code    incident_datetime      report_datetime  \
0    571905  2019-01-01 00:00:00  2019-01-02 06:11:00   
1   3790870  2019-01-01 00:03:00  2020-02-20 13:48:00   
2    599075  2019-01-01 00:05:00  2019-01-01 10:35:00   
3    571924  2019-01-01 00:20:00  2019-01-02 05:12:00   
4    599523  2019-01-01 00:25:00  2019-01-04 09:42:00   

                                     province_th          province_en  \
0                             ลพบุรี               Loburi   
1              อุบลราชธานี     Ubon Ratchathani   
2  ประจวบคีรีขันธ์  Prachuap Khiri Khan   
3                    เชียงใหม่           Chiang Mai   
4                    นครสวรรค์         Nakhon Sawan   

                      agency  \
0  department of rural roads   
1     department of highways   
2     department of highways   
3  department of rural roads   
4     department of highways   

                                               route           vehicle_type  \
0  แยกทางหลวงหมายเà...             motorcycle   
1  เดชอุดม - อุบลราชà...  private/passenger car   
2  ปราณบุรี - ปากน้ำà...             motorcycle   
3  เชื่อมทางหลวงหมà...             motorcycle   
4         เกยไชย - ศรีมงคล  private/passenger car   

                                  presumed_cause  \
0         driving under the influence of alcohol   
1                                       speeding   
2                                       speeding   
3         driving under the influence of alcohol   
4  cutting in closely by people/vehicles/animals   

                        accident_type  number_of_vehicles_involved  \
0                               other                            1   
1    rollover/fallen on straight road                            1   
2  head-on collision (not overtaking)                            2   
3                               other                            1   
4    rollover/fallen on straight road                            1   

   number_of_fatalities  number_of_injuries weather_condition   latitude  \
0                     0                   2             clear  14.959105   
1                     0                   2             clear  15.210738   
2                     1                   0             clear  12.374259   
3                     0                   1             clear  18.601721   
4                     0                   0             clear  15.866389   

    longitude road_description slope_description  
0  100.873463    straight road          no slope  
1  104.862689    straight road          no slope  
2   99.907949       wide curve        slope area  
3   98.804204    straight road          no slope  
4  100.590010    straight road          no slope

We have imported our data, so we can check the size of the dataset

In [ ]:
data.shape

(81735, 18)

We have 81735 row and 18 columns in the dataset

In [ ]:
data.columns

Index(['acc_code', 'incident_datetime', 'report_datetime', 'province_th',
       'province_en', 'agency', 'route', 'vehicle_type', 'presumed_cause',
       'accident_type', 'number_of_vehicles_involved', 'number_of_fatalities',
       'number_of_injuries', 'weather_condition', 'latitude', 'longitude',
       'road_description', 'slope_description'],
      dtype='object')

we generated the list of columns available, but out first column is not properly named..

So we have successfully renamed the column

Generate the information of the dataset

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 81735 entries, 0 to 81734
Data columns (total 18 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   acc_code                     81735 non-null  int64  
 1   incident_datetime            81735 non-null  object 
 2   report_datetime              81735 non-null  object 
 3   province_th                  81735 non-null  object 
 4   province_en                  81735 non-null  object 
 5   agency                       81735 non-null  object 
 6   route                        81735 non-null  object 
 7   vehicle_type                 81735 non-null  object 
 8   presumed_cause               81735 non-null  object 
 9   accident_type                81735 non-null  object 
 10  number_of_vehicles_involved  81735 non-null  int64  
 11  number_of_fatalities         81735 non-null  int64  
 12  number_of_injuries           81735 non-null  int64  
 13  weather_conditio

We can observe that we have different data types such ad float, int, object

In [ ]:
# code to generate the number of null values present in the columns
data.isnull().sum()

acc_code                         0
incident_datetime                0
report_datetime                  0
province_th                      0
province_en                      0
agency                           0
route                            0
vehicle_type                     0
presumed_cause                   0
accident_type                    0
number_of_vehicles_involved      0
number_of_fatalities             0
number_of_injuries               0
weather_condition                0
latitude                       359
longitude                      359
road_description                 0
slope_description                0
dtype: int64

we can observe that column latitude and longitude contain the same number of missing values

## **DATA CLEANING**

### **Incident_datetime column**

In [ ]:
# We start by analysing the 'incident datetime' column , we convert the type of column into a datetime
#data['incident_datetime'] = pd.to_datetime(data['incident_datetime'])

In [ ]:
data['incident_datetime'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 81735 entries, 0 to 81734
Series name: incident_datetime
Non-Null Count  Dtype 
--------------  ----- 
81735 non-null  object
dtypes: object(1)
memory usage: 638.7+ KB


The column has been converted to a datetime column. so lets check if there are no null values in the dataset

In [ ]:
data['incident_datetime'].isnull().sum()

0

We can confirm that there are no null values.

We can sort the data set with the incident datetime column

In [ ]:
data.sort_values(by='incident_datetime', inplace=True)

Now we can be sure that the dataset has been sorted

**Now we want to create two different column from the datetime column, one of the date (Year-Month-Day) and another for the time zone (Hour-minute-Seconds)**

In [ ]:
# Creating a new column of just the date
# Creating a new column of just the time
data['incident_date'] = data['incident_datetime'].str[:10]
data['incident_time'] = data['incident_datetime'].str[10:]


Now we can drop/del the incident_datetime column and also rearrange the data bring the newly created columns to the front

In [ ]:

# we use the pop function
first_col = data.pop('incident_date')
second_col = data.pop('incident_time')

# then we insert it the column number position, the name of the columns, and the coloumn we popped
data.insert(1, 'Incident_date', first_col)
data.insert(2, 'Incident_time', second_col)


### **Report_datetime column**

Now we want to create two different column from the datetime column, one of the date (Year-Month-Day) and another for the time zone (Hour-minute-Seconds)

In [ ]:
# Creating a new column of just the date
# Creating a new column of just the time
data['report_date'] = data['report_datetime'].str[:10]
data['report_time'] = data['report_datetime'].str[10:]

Now we can drop/del the incident_datetime column and also rearrange the data bring the newly created columns to the front

i use Ctrl + shift + L to multiple select a repeated string after highlighting of of it

In [ ]:
# we use the pop function
first_col = data.pop('report_date')
second_col = data.pop('report_time')

# then we insert it the column number position, the name of the columns, and the coloumn we popped
data.insert(3, 'report_date', first_col)
data.insert(4, 'report_time', second_col)



In [ ]:
# we can use the drop function to delete the report datetime column
data.drop(['report_datetime','incident_datetime'], axis=1, inplace=True)
data.head()


acc_code Incident_date Incident_time report_date report_time  \
0    571905    2019-01-01      00:00:00  2019-01-02    06:11:00   
1   3790870    2019-01-01      00:03:00  2020-02-20    13:48:00   
2    599075    2019-01-01      00:05:00  2019-01-01    10:35:00   
3    571924    2019-01-01      00:20:00  2019-01-02    05:12:00   
4    599523    2019-01-01      00:25:00  2019-01-04    09:42:00   

                                     province_th          province_en  \
0                             ลพบุรี               Loburi   
1              อุบลราชธานี     Ubon Ratchathani   
2  ประจวบคีรีขันธ์  Prachuap Khiri Khan   
3                    เชียงใหม่           Chiang Mai   
4                    นครสวรรค์         Nakhon Sawan   

                      agency  \
0  department of rural roads   
1     department of highways   
2     department of highways   
3  department of rural roads   
4     department of highways   

                                               route           vehicle_type  \
0  แยกทางหลวงหมายเà...             motorcycle   
1  เดชอุดม - อุบลราชà...  private/passenger car   
2  ปราณบุรี - ปากน้ำà...             motorcycle   
3  เชื่อมทางหลวงหมà...             motorcycle   
4         เกยไชย - ศรีมงคล  private/passenger car   

                                  presumed_cause  \
0         driving under the influence of alcohol   
1                                       speeding   
2                                       speeding   
3         driving under the influence of alcohol   
4  cutting in closely by people/vehicles/animals   

                        accident_type  number_of_vehicles_involved  \
0                               other                            1   
1    rollover/fallen on straight road                            1   
2  head-on collision (not overtaking)                            2   
3                               other                            1   
4    rollover/fallen on straight road                            1   

   number_of_fatalities  number_of_injuries weather_condition   latitude  \
0                     0                   2             clear  14.959105   
1                     0                   2             clear  15.210738   
2                     1                   0             clear  12.374259   
3                     0                   1             clear  18.601721   
4                     0                   0             clear  15.866389   

    longitude road_description slope_description  
0  100.873463    straight road          no slope  
1  104.862689    straight road          no slope  
2   99.907949       wide curve        slope area  
3   98.804204    straight road          no slope  
4  100.590010    straight road          no slope

**From our observation of the datetime columns, we noticed that in some rows there are difference in the reported  date/time and when the inccident occurred.... we can get some insight by analysing this as to why the delay occurred**

### **Province_th Column**

In [ ]:
# preview the column
data['province_th']

0                                   ลพบุรี
1                    อุบลราชธานี
2        ประจวบคีรีขันธ์
3                          เชียงใหม่
4                          นครสวรรค์
                             ...                      
81730                            นนทบุรี
81731                                     ตราด
81732                      มหาสารคาม
81733                      กำแพงเพชร
81734                         เชียงราย
Name: province_th, Length: 81735, dtype: object

From the description, this column is the names of the provinces the incidents occurred..

I decided to drop to column, the texts were not well represented..

In [ ]:
data.drop('province_th', axis=1, inplace=True)

### **Province_en Column**

In [ ]:
#preview the column
data['province_en']

0                     Loburi
1           Ubon Ratchathani
2        Prachuap Khiri Khan
3                 Chiang Mai
4               Nakhon Sawan
                ...         
81730             Nonthaburi
81731                   Trat
81732          Maha Sarakham
81733         Kamphaeng Phet
81734             Chiang Rai
Name: province_en, Length: 81735, dtype: object

Lets check the number of province involved

In [ ]:
# LIst of the unique province in the dataset
data['province_en'].unique()

array(['Loburi', 'Ubon Ratchathani', 'Prachuap Khiri Khan', 'Chiang Mai',
       'Nakhon Sawan', 'Mae Hong Son', 'Chumphon', 'Sing Buri',
       'Songkhla', 'Lamphun', 'Trat', 'Phuket', 'Saraburi', 'Ratchaburi',
       'Phra Nakhon Si Ayutthaya', 'Nakhon Ratchasima',
       'Nakhon Si Thammarat', 'Chaiyaphum', 'Kalasin', 'Suphan Buri',
       'Phetchaburi', 'Phrae', 'Chai Nat', 'Prachin Buri',
       'Nakhon Pathom', 'Kanchanaburi', 'Phetchabun', 'Ang Thong',
       'Nonthaburi', 'Samut Prakan', 'Bangkok', 'Phayao', 'Phatthalung',
       'Yala', 'Maha Sarakham', 'Surat Thani', 'Amnat Charoen',
       'Nong Khai', 'Nan', 'Phangnga', 'Narathiwat', 'Samut Sakhon',
       'Chanthaburi', 'Samut Songkhram', 'Phitsanulok', 'Pathum Thani',
       'Tak', 'Loei', 'Chiang Rai', 'Chachoengsao', 'Buri Ram',
       'Uthai Thani', 'Krabi', 'Surin', 'Udon Thani', 'Si Sa Ket',
       'Uttaradit', 'Khon Kaen', 'Kamphaeng Phet', 'Yasothon', 'Satun',
       'Nakhon Nayok', 'Chon Buri', 'Rayong', 'buogkan'

Let convert this column to a categorical column

In [ ]:
data['province_en'] = data['province_en'].astype('category')

In [ ]:
data['province_en'].info()

<class 'pandas.core.series.Series'>
Int64Index: 81735 entries, 0 to 81734
Series name: province_en
Non-Null Count  Dtype   
--------------  -----   
81735 non-null  category
dtypes: category(1)
memory usage: 721.0 KB


In [ ]:
data.head()

acc_code Incident_date Incident_time report_date report_time  \
0    571905    2019-01-01      00:00:00  2019-01-02    06:11:00   
1   3790870    2019-01-01      00:03:00  2020-02-20    13:48:00   
2    599075    2019-01-01      00:05:00  2019-01-01    10:35:00   
3    571924    2019-01-01      00:20:00  2019-01-02    05:12:00   
4    599523    2019-01-01      00:25:00  2019-01-04    09:42:00   

           province_en                     agency  \
0               Loburi  department of rural roads   
1     Ubon Ratchathani     department of highways   
2  Prachuap Khiri Khan     department of highways   
3           Chiang Mai  department of rural roads   
4         Nakhon Sawan     department of highways   

                                               route           vehicle_type  \
0  แยกทางหลวงหมายเà...             motorcycle   
1  เดชอุดม - อุบลราชà...  private/passenger car   
2  ปราณบุรี - ปากน้ำà...             motorcycle   
3  เชื่อมทางหลวงหมà...             motorcycle   
4         เกยไชย - ศรีมงคล  private/passenger car   

                                  presumed_cause  \
0         driving under the influence of alcohol   
1                                       speeding   
2                                       speeding   
3         driving under the influence of alcohol   
4  cutting in closely by people/vehicles/animals   

                        accident_type  number_of_vehicles_involved  \
0                               other                            1   
1    rollover/fallen on straight road                            1   
2  head-on collision (not overtaking)                            2   
3                               other                            1   
4    rollover/fallen on straight road                            1   

   number_of_fatalities  number_of_injuries weather_condition   latitude  \
0                     0                   2             clear  14.959105   
1                     0                   2             clear  15.210738   
2                     1                   0             clear  12.374259   
3                     0                   1             clear  18.601721   
4                     0                   0             clear  15.866389   

    longitude road_description slope_description  
0  100.873463    straight road          no slope  
1  104.862689    straight road          no slope  
2   99.907949       wide curve        slope area  
3   98.804204    straight road          no slope  
4  100.590010    straight road          no slope

### **Agency Column**

In [ ]:
## Getting the categories of the column
data['agency'].unique()

array(['department of rural roads', 'department of highways',
       'expressway authority of thailand'], dtype=object)

We have three government agency involved :

*   department of rural roads  
*   department of highways
*   expressway authority of thailand




### **Route Column**

In [ ]:
data.route.head()

0    แยกทางหลวงหมายเà...
1    เดชอุดม - อุบลราชà...
2    ปราณบุรี - ปากน้ำà...
3    เชื่อมทางหลวงหมà...
4           เกยไชย - ศรีมงคล
Name: route, dtype: object

Also the texts in this column were not well represented so i drop/del the column

In [ ]:
data.drop('route', axis=1, inplace=True)

In [ ]:
data.head()

acc_code Incident_date Incident_time report_date report_time  \
0    571905    2019-01-01      00:00:00  2019-01-02    06:11:00   
1   3790870    2019-01-01      00:03:00  2020-02-20    13:48:00   
2    599075    2019-01-01      00:05:00  2019-01-01    10:35:00   
3    571924    2019-01-01      00:20:00  2019-01-02    05:12:00   
4    599523    2019-01-01      00:25:00  2019-01-04    09:42:00   

           province_en                     agency           vehicle_type  \
0               Loburi  department of rural roads             motorcycle   
1     Ubon Ratchathani     department of highways  private/passenger car   
2  Prachuap Khiri Khan     department of highways             motorcycle   
3           Chiang Mai  department of rural roads             motorcycle   
4         Nakhon Sawan     department of highways  private/passenger car   

                                  presumed_cause  \
0         driving under the influence of alcohol   
1                                       speeding   
2                                       speeding   
3         driving under the influence of alcohol   
4  cutting in closely by people/vehicles/animals   

                        accident_type  number_of_vehicles_involved  \
0                               other                            1   
1    rollover/fallen on straight road                            1   
2  head-on collision (not overtaking)                            2   
3                               other                            1   
4    rollover/fallen on straight road                            1   

   number_of_fatalities  number_of_injuries weather_condition   latitude  \
0                     0                   2             clear  14.959105   
1                     0                   2             clear  15.210738   
2                     1                   0             clear  12.374259   
3                     0                   1             clear  18.601721   
4                     0                   0             clear  15.866389   

    longitude road_description slope_description  
0  100.873463    straight road          no slope  
1  104.862689    straight road          no slope  
2   99.907949       wide curve        slope area  
3   98.804204    straight road          no slope  
4  100.590010    straight road          no slope

### **Vehicle Type Column**

In [ ]:
data['vehicle_type'].head()

0               motorcycle
1    private/passenger car
2               motorcycle
3               motorcycle
4    private/passenger car
Name: vehicle_type, dtype: object

In [ ]:
## checking if there is no null value
data['vehicle_type'].isnull().sum()

0

we have no null values

### **Presumed Cause Column**

In [ ]:
data['presumed_cause'].head()

0           driving under the influence of alcohol
1                                         speeding
2                                         speeding
3           driving under the influence of alcohol
4    cutting in closely by people/vehicles/animals
Name: presumed_cause, dtype: object

## **DATA ANALYSIS AND EXPLORATION**



## **DATA VISUALIZATION**